Extensions/TokenAcquisition.cs

```diff
@@ -0,0 +1,188 @@
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Client;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using TodoListService.Extensions;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    /// <summary>
+    /// Extension class enabling adding the TokenAcquisition service
+    /// </summary>
+    public static class TokenAcquisitionExtension
+    {
+        /// <summary>
+        /// Add the token acquisition service.
+        /// </summary>
+        /// <param name="services">Service collection</param>
+        /// <returns>the service collection</returns>
+        public static IServiceCollection AddTokenAcquisition(this IServiceCollection services)
+        {
+            // Token acquisition service
+            services.Configure<AzureAdOptions>(myoptions => { });
+            services.AddTransient<ITokenAcquisition, TokenAcquisition>();
+            return services;
+        }
+    }
+
+    /// <summary>
+    /// Token acquisition service
+    /// </summary>
+    public class TokenAcquisition : ITokenAcquisition
+    {
+        /// <summary>
+        /// Constructor of the TokenAcquisition service. This requires the Azure AD Options to 
+        /// configure the confidential client application
+        /// </summary>
+        /// <param name="options">Options to configure the application</param>
+        public TokenAcquisition(IOptions<AzureAdOptions> options)
+        {
+            azureAdOptions = options.Value;
+        }
+
+        private AzureAdOptions azureAdOptions;
+
+        /// <summary>
+        /// The goal of this method is, when a user is authenticated, to add the user's account in the MSAL.NET cache
+        /// so that this token can then be used to acquire a token on On-behalf-of the user befor call to downstream APIs.
+        /// </summary>
+        /// <param name="userAccessToken">Access token used to call this Web API</param>
+        /// <example>
+        /// From configuration of the Authentication of the ASP.NET Core Web API 
+        /// <code>JwtBearerOptions option;</code>
+        /// 
+        /// Subscribe to the token validated event
+        /// <code>
+        /// options.Events = new JwtBearerEvents();
+        /// options.Events.OnTokenValidated = OnTokenValidated;
+        /// }
+        /// </code>
+        /// 
+        /// And then in the OnTokenValidated method:
+        /// <code>
+        /// private async Task OnTokenValidated(TokenValidatedContext context)
+        /// {
+        ///  JwtSecurityToken accessToken = context.SecurityToken as JwtSecurityToken;
+        ///  _tokenAcquisition.AddAccountToCache(accessToken);
+        /// }
+        /// </code>
+        /// </example>
+        public void AddAccountToCacheFromJwt(JwtSecurityToken jwtToken, IEnumerable<string> scopes)
+        {
+            string userAccessTokenForThisApi = jwtToken.RawData;
```

Comment:
remove variable and user RawData below in UserAssertion call.

Reply:
why?

---
In reply to: [212387951](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/7#discussion_r212387951) [](ancestors = 212387951)

Reply:
Made this change. Since the variable is only used in one location I suppose it isn't really necessary.